### ISSUE TITLE
glowroot - sed: couldn't open temporary file ./sedShvrpf: Permission denied

### ISSUE THREAD

**Issue by robinroos** [novumrgi/helm#10]
I'm using the NOVUM-RGI Helm Chart (v0.13.7) to deploy Glowroot (Central Collector) into OpenShift 4.6.

https://artifacthub.io/packages/helm/novum-rgi-charts/glowroot

Upon attempting to start a Pod of the Deployment, the Pod goes into state CrashLoopBackOff with the log containing only the following line:

`sed: couldn't open temporary file ./sedShvrpf: Permission denied`

So far everything else has gone really well. I merely had to provide runAsUser and fsGroup values for the StatefulSet, and provision the Persistent Volume Claim. The StatefulSet is now started.

Now it would seem I am stuck at the last step!

**Comment by luhahn** [comment]
Never seen the sed error before. Maybe this is solved by using the correct fsGroup.

**Comment by robinroos** [comment]
I have used Helm to uninstall 1.0.4 and to install 1.0.5 using the following .yml:

```
securityContext:
  runAsUser: 1000550000
  runAsGroup: 1000550000
  fsGroup: 1000550000
```

Upon attempting to start the Deployment pod "glowroot" I get the error:

`sed: couldn't open temporary file ./sed7eVRRG: Permission denied`

Since the sed invocation does not form part of the YAML for the Pod, perhaps it is part of the Container novumrgi/glowroot-central 0.13.7 ?

Knowing where the sed invocation is situated will help us to resolve the issue.

**Comment by robinroos** [comment]
@luhahn I believe you might be able to reproduce this yourself by creating a new namespace with the following annotations:

```
Annotations:            k8s.ovn.org/multicast-enabled=true
                        openshift.io/description=
                        openshift.io/display-name=
                        openshift.io/requester=clusteradmin
                        openshift.io/sa.scc.mcs=s0:c23,c22
                        openshift.io/sa.scc.supplemental-groups=1000550000/10000
                        openshift.io/sa.scc.uid-range=1000550000/10000
```

and then attempting to install Glowroot with a glowroot.yml file containing:

```
securityContext:
  runAsUser: 1000550000
  runAsGroup: 1000550000
  fsGroup: 1000550000
```

The specific Helm command I used was:

`helm install glowroot -f glowroot.yml novum-rgi-charts/glowroot --version 1.0.5 `


Note that the Stateful Set (a product of the Bitnami helm chart) will need to be manually updated with 100055000 for runAsUser and fsGroup before it can be started.

The error

`sed: couldn't open temporary file ./sed7eVRRG: Permission denied`

Appears as the only log entry of the Pod of the Deployment "glowroot".

Hopefully you will be able to reproduce this.

**Comment by luhahn** [comment]
Can you tell me what permissions the folder /usr/share/glowroot-central has for you (in your container)?

**Comment by robinroos** [comment]
@luhahn , I don't think I know how to establish that.

From Deployment "glowroot" I can go to Pod "glowroot-6c6654f8f5-rb9s2".  Log shows the sed error.  Terminal lists one container "glowroot" and says "The terminal connection has closed". 

From the Pod Details I can go to the Container "glowroot".

Amongst the panels on that page, Container Details, Image Details, Network, Ports, Mounted Volumes (just Secrets), and Environment Variables, there is nothing about file system level permissions.

I presume I should have to get a shell session on the running Pod to answer your question, but the Pod is not running at all or is not running for long enough for me to get a Terminal session.

I'm doing all of this with the OpenShift "Administrator" profile selected.  Perhaps there are other tools available if I change to the OpenShift Developer profile?

Thanks, Robin.

**Comment by luhahn** [comment]
To be honest I'm not familiar with OpenShift at all :/. The problem is, that I'm not really a developer of glowroot, so I don't have an idea where exactly sed is executed.

Since the log says ./ I'm guessing it is  /usr/share/glowroot-central because the binary is located there

**Comment by robinroos** [comment]
Lucas, I have cross-raised this at https://github.com/glowroot/glowroot/issues/767

I note that file https://github.com/glowroot/glowroot/blob/master/central/Dockerfile , which contains a `sed` invocation, also contains:

`USER glowroot:glowroot`

Is it possible in OpenShift to configure mappings between group "glowroot" and user "glowroot" to 1000550000 ?

**Comment by robinroos** [comment]
Do you still build your own NOVUM-RGI image of Glowroot Central?

I suspect that for this to run within OpenShift I would need, in the Dockerfile, something akin to

```
chown a+w /usr/share/glowroot-central
ls -ld /usr/share/glowroot-central
```

before the `sed` invocation.  The second command would at least generate some log output regarding permissions of that folder.

**Comment by luhahn** [comment]
Yes we do. Mh I'm somehow not convinced, that this would be the right way to handle that permissions. Sorry btw for the late replies. Has been kind of a rough first week after vacation. I will check how to handle those permissions correctly in kubernetes tomorrow.

**Comment by robinroos** [comment]
Just wondering, do I need to define a Service Account "glowroot" ?

**Comment by luhahn** [comment]
It appears, that this could be a solution. I will add the service account values to the chart. 

To be honest, I've never needed a service account and I'm not familiar with it :D

**Comment by robinroos** [comment]
After perusing much OpenShift documentation it appears that a Service Account approach would not work.

Instead, the following are recommended to be included in the DOCKERFILE before sed is invoked:

```
RUN adduser --disabled-password --gid 0 --gecos "glowroot user" glowroot
USER glowroot
RUN chown glowroot:root /home/glowroot && chmod 0775 /home/glowroot
ENV HOME /home/glowroot
```
The use of `--gid 0` is crucial.  The user OpenShift uses to run the Container will not be "glowroot", but it will be a user in the "root" group.

Is this something you would be able to do in your Glowroot container?

**Comment by robinroos** [comment]
@luhahn, can we discuss this approach as a possible solution?  I'm really keen to get OpenShift deployment working.

**Comment by robinroos** [comment]
I trimmed down the suggested enhancement to merely creating the user "glowroot" with gid 0.

I guess an alternative would be to use a chgrp invocation to assign gid 0 across the folder structures after the unzip, before the sed invocation.  That approach might work better for environments where the user "glowroot" is already defined (and addurer might therefore fail).

**Comment by robinroos** [comment]
`chgrp -R +0 .`

should do it.

**Comment by luhahn** [comment]
Sorry for the late answer again. I have a lot to do including some urgent deadlines.

To your question. I built the glowroot image, yes. So, making some changes would be possible. 
However I'm still not convinced that this is a good solution, since the most docker containers are using user 1000. I cannot imagine, that this is the way to go.

But I want, that you can use the glowroot chart. 

So how about some extra init parameters in the helm chart, where you can inject all the code you want to initialize the container?

**Comment by luhahn** [comment]
@robinroos Can you test glowroot chart version 1.0.6?

I added a parameter:

```yaml
glowroot:
  extraInitConfig: |
    chmod a+rwx /usr/share/glowroot-central
```

You should be able to modify the glowroot-central folders permissions there at least my tests where successful

**Comment by robinroos** [comment]
Thanks @luhahn .

Using

`>helm install glowroot -f glowroot.yml novum-rgi-charts/glowroot --version 1.0.6`

with

```
securityContext:
  runAsUser: 1000550000
  runAsGroup: 1000550000
  fsGroup: 1000550000
glowroot:
  extraInitConfig: |
    ls -lsd /usr/share/glowroot-central
```

I see only

`sed: couldn't open temporary file ./sedzFdM1u: Permission denied`

Is the `extraInitConfig` executed after the zip extract but before the `sed` command?  If it is executed before the zip extract then I cannot change folder permissions, and if it is executed after `sed` then I cannot avoid that failure.

**Comment by luhahn** [comment]
I think you're refering to this part of the dockerfile:

```
RUN unzip -d /usr/share /tmp/glowroot-central.zip \
    && rm /tmp/glowroot-central.zip \
    && sed -i 's/^cassandra.contactPoints=$/cassandra.contactPoints=cassandra/' /usr/share/glowroot-central/glowroot-central.properties \
    && echo '\ncassandra.symmetricEncryptionKey=' >> /usr/share/glowroot-central/glowroot-central.properties \
    && groupadd -r glowroot \
    && useradd --no-log-init -r -g glowroot glowroot \
    && chown -R glowroot:glowroot /usr/share/glowroot-central \
    && chmod a+x /usr/local/bin/docker-entrypoint.sh \
    && chmod a+x /usr/local/bin/glowroot-central.sh
```

All the parts until CMD are not executed in the helm file. This part is only run on docker build and is already executed in the image before uploading it to docker hub. So this can't be the part where it fails.

I suggest, that you add 
```
chmod a+rwx /usr/share/glowroot-central
```
To your glowroot.extraInitConfig

**Comment by robinroos** [comment]
Yes, that's what I did first.  I reverted to `ls -lsd /usr/share/glowroot-central` only once I saw the sed error again.

Can we try with an image built using:

```
RUN echo unzipping /tmp/glowroot-central.zip
    && unzip -d /usr/share /tmp/glowroot-central.zip \
    && rm /tmp/glowroot-central.zip \
    && useradd --no-log-init -r -g root glowroot \
    && chown -R glowroot:root /usr/share/glowroot-central \
    && chmod 664 /usr/share/glowroot-central \
    && chmod a+x /usr/local/bin/docker-entrypoint.sh \
    && chmod a+x /usr/local/bin/glowroot-central.sh \
    && sed -i 's/^cassandra.contactPoints=$/cassandra.contactPoints=cassandra/' /usr/share/glowroot-central/glowroot-central.properties \
    && echo '\ncassandra.symmetricEncryptionKey=' >> /usr/share/glowroot-central/glowroot-central.properties 
```

I use group root instead of group glowroot (so I removed grpadd).
I set 664 across the filesystem.
I invoke sed at the end.

This should result in an image upon which anything user "glowroot" can do, can also be done by any member of the "root" group.

OpenShift will run your Container with anarbitrary user (not "glowroot") who is a member of the "root" group.

Chances are that change would result in an image which works for OpenShift, and is not deficient for other container platforms.

I moved sed to the end just in case.  I note your assertion that sed is only involved in building the image, but I do not know the OpenShift deployment details in sufficient detail to be able to concurr and there seems to be no disadvantage in doing so.  

I also added an echo.  That will tell us once and for all whether OpenShift is performing the whole unzip and set permissions cycle every time the Pod (Container) is started.  The echo can be removed once we show the filesystem permissions, based on group root, are appropriate.

**Comment by luhahn** [comment]
Just uploaded the new image. Please use image.tag: openshift in your values

**Comment by luhahn** [comment]
Still not convinced that this is the way to go in openshift. This would mean, that every docker image you're using needs to be adjustet.

**Comment by robinroos** [comment]
Ok, deployed 1.0.6 with image.tag: openshift

The following error is in the log:

`sed: can't read glowroot-central.properties: Permission denied`

So this is DIFFERENT, which is GOOD!

**Comment by robinroos** [comment]
So:

1. OpenShift is executing that part of the Dockerfile.
2. We need to add -R to the chown, e.g.
`&& chmod -R 664 /usr/share/glowroot-central \`

**Comment by luhahn** [comment]
You mean, that you saw "unzipping /tmp/glowroot-central.zip" in your logs?

**Comment by luhahn** [comment]
"OpenShift is executing that part of the Dockerfile."

I would rather say docker build executed that part of the dockerfile, which set the permissions so openshift was able to use the directory.

**Comment by robinroos** [comment]
> You mean, that you saw "unzipping /tmp/glowroot-central.zip" in your logs?

No, I didn't.  So all I can say is that the permissions on glowroot-central.properties are not yet correct.  Of course, the fact that the error was from a `sed` invocation is intriguing.  Perhaps only stderr from the shell is routed to the log window - echo will write to stdout.

**Comment by robinroos** [comment]
`echo unzipping /tmp/glowroot-central.zip >&2`

?

**Comment by luhahn** [comment]
Okay, so Im currently pushing the chmod with -R option. 

Please note that these changes are not inside a PR in glowroot master. This is my own fork, which I created to get the helm chart to work. 

Once my PR is merged into glowroot master you won't be able to receive any updates with the official glowroot image. 
(https://github.com/glowroot/glowroot/pull/733)

I suggest, that you find a way to do this correctly on openshift. I can't imagine, that everyone is changing images just to get it to run in openshift

**Comment by luhahn** [comment]
Regarding the echo unzipping. This won't show up in your logs, because this is not executed in the image. It is pre build in the docker image. You won't see that.

**Comment by robinroos** [comment]
Still v 1.0.6 with image.tag: openshift ?

**Comment by luhahn** [comment]
yes

**Comment by robinroos** [comment]
Same issue even with `helm repo update` before `helm install`.

Can you drop the full Dockerfile extract here for my review?

Thanks.

**Comment by luhahn** [comment]
helm repo update only updates the helm chart. This was a docker image update. 
Please make sure, your imagePullPolicy is set to Always.

in addition you could just docker pull the image on your node

```
RUN echo unzipping /tmp/glowroot-central.zip \
    && unzip -d /usr/share /tmp/glowroot-central.zip \
    && rm /tmp/glowroot-central.zip \
    && useradd --no-log-init -r -g root glowroot \
    && chown -R glowroot:root /usr/share/glowroot-central \
    && chmod -R 664 /usr/share/glowroot-central \
    && chmod a+x /usr/local/bin/docker-entrypoint.sh \
    && chmod a+x /usr/local/bin/glowroot-central.sh \
    && sed -i 's/^cassandra.contactPoints=$/cassandra.contactPoints=cassandra/' /usr/share/glowroot-central/glowroot-central.properties \
    && echo '\ncassandra.symmetricEncryptionKey=' >> /usr/share/glowroot-central/glowroot-central.properties 
```

**Comment by robinroos** [comment]
That chmod -R is dangerous as it removed x (search) permission from directories.

I have achieved very limited $ access to a Pod, and I observe:

`drw-rw-r--.  1 glowroot root   32 Jan 26 16:26 glowroot-central`

**Comment by robinroos** [comment]
Can we go with `chmod -R 774` just for a brief while?  I will then be able to perform other `pwd` and `ls` commands on the pod.

**Comment by robinroos** [comment]
Note, 775, so that all users also have x (search) permission.

**Comment by robinroos** [comment]
Not sure 775 is right.  774 feels better.

**Comment by robinroos** [comment]
Just touched a file to check, and my user is a member of the root group:

```
$ touch /tmp/robin
$ ls -l /tmp/robin
-rw-r--r--. 1 1000550000 root 0 Jan 26 16:42 /tmp/robin
$
```

**Comment by robinroos** [comment]
The alternative is to replace 664 with two g+rw, which is the actual desired effect, relying on the fact that g probably already has x permission on directories.

So `&& chmod -R 664 /usr/share/glowroot-central \` becomes:

`&& chmod -R g+rw /usr/share/glowroot-central \`

**Comment by luhahn** [comment]
I had some time thinking about this issue, and i might have an idea how to fix this the right way without modifying the helm chart. I will prepare a new chart version tomorrow. And also update the openshift image once again.

Will notice you once I'am finished.

**Comment by luhahn** [comment]
updated the image with your requested changes.

The problem you're having could be solved if I mount a persistent volume to /usr/share/glowroot-central, which would be mounted with your security context. However since the dockerfile is already populating this directory I need to search for a good solution here.
I want to avoid to copy the contents on every launch of the image.

**Comment by robinroos** [comment]
Eurika!

```
0 drwxrwxr-x. 1 glowroot root 67 Jan 27 08:50 /usr/share/glowroot-central
2021-01-27 08:50:47.606 INFO  org.glowroot - Glowroot version: 0.13.7-SNAPSHOT, built 2020-06-29 08:16:09 +0000
2021-01-27 08:50:47.609 INFO  org.glowroot - Java version: 1.8.0_282
2021-01-27 08:50:47.780 INFO  c.datastax.driver.core.ClockFactory - Using native clock to generate timestamps.
2021-01-27 08:50:47.787 INFO  com.datastax.driver.core - DataStax Java driver 3.7.2 for Apache Cassandra
2021-01-27 08:50:47.790 INFO  c.d.driver.core.GuavaCompatibility - Detected Guava >= 19 in the classpath, using modern compatibility layer
2021-01-27 08:50:47.889 INFO  com.datastax.driver.core.NettyUtil - Did not find Netty's native epoll transport in the classpath, defaulting to NIO.
2021-01-27 08:50:48.499 INFO  c.d.d.c.p.DCAwareRoundRobinPolicy - Using data-center name 'datacenter1' for DCAwareRoundRobinPolicy (if this is incorrect, please provide the correct datacenter name with DCAwareRoundRobinPolicy constructor)
2021-01-27 08:50:48.500 INFO  com.datastax.driver.core.Cluster - New Cassandra host glowroot-cassandra/172.30.113.24:9042 added
2021-01-27 08:50:49.256 INFO  org.glowroot - connected to Cassandra (version 3.11.9), using keyspace 'glowroot' (replication factor 1) and consistency level QUORUM
2021-01-27 08:50:49.626 INFO  org.glowroot - creating glowroot central schema...
2021-01-27 08:51:19.927 INFO  org.glowroot - glowroot central schema created
2021-01-27 08:51:20.036 INFO  org.glowroot - gRPC listening on 0.0.0.0:8181
2021-01-27 08:51:20.460 INFO  org.glowroot - UI listening on 0.0.0.0:4000
2021-01-27 08:51:20.460 INFO  org.glowroot - startup complete
```

So, all that was needed was to use group root instead of group glowroot, and to set filesystem permissions the same for the user and group.

**Comment by robinroos** [comment]
I will now remove my extraInitConfig.

If you can provide me with the WHOLE dockerfile for the openshift  image tag, then I will raise a PR over at glowroot to have their dockerfile made openshift-compatible.

**Comment by robinroos** [comment]
@luhahn please provide me with the full Dockerfile from the openshift image tag so that I can raise the necessary PR against the glowroot repo.

Thanks, Robin.

**Comment by luhahn** [comment]
Not sure if this would be accepted, since the chmod stuff is still not nice. Would be better to deploy the glowroot contents after the persistent volume is mounted. Which should be done in the entrypoint.sh

```
FROM openjdk:8

COPY target/glowroot-central-*.zip /tmp/glowroot-central.zip
COPY docker-entrypoint.sh /usr/local/bin/
COPY glowroot-central.sh /usr/local/bin/

RUN echo unzipping /tmp/glowroot-central.zip \
    && unzip -d /usr/share /tmp/glowroot-central.zip \
    && rm /tmp/glowroot-central.zip \
    && useradd --no-log-init -r -g root glowroot \
    && chown -R glowroot:root /usr/share/glowroot-central \
    && chmod -R g+rw /usr/share/glowroot-central \
    && chmod a+x /usr/local/bin/docker-entrypoint.sh \
    && chmod a+x /usr/local/bin/glowroot-central.sh \
    && sed -i 's/^cassandra.contactPoints=$/cassandra.contactPoints=cassandra/' /usr/share/glowroot-central/glowroot-central.properties \
    && echo '\ncassandra.symmetricEncryptionKey=' >> /usr/share/glowroot-central/glowroot-central.properties 

EXPOSE 4000 8181

RUN touch /usr/sbin/additional-config.sh && \
    chmod a+x /usr/sbin/additional-config.sh

WORKDIR /usr/share/glowroot-central

USER glowroot:glowroot

ENV GLOWROOT_OPTS -XX:+UnlockExperimentalVMOptions -XX:+UseCGroupMemoryLimitForHeap

ENTRYPOINT ["docker-entrypoint.sh"]

CMD additional-config.sh && glowroot-central.sh
```
